Core/src/Geometry/DetectorVolume.cpp

```diff
@@ -0,0 +1,228 @@
+  }
+  // Check exclusion through subvolume
+  for (const auto v : volumes()) {
+    if (v->inside(gctx, position)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void Acts::Experimental::DetectorVolume::updateNavigationState(
+    const GeometryContext& gctx, NavigationState& nState) const {
+  // This volume still has sub volumes, it needs a double-check
+  // if the internal volumes can be excluded.
+  //
+  // This can be avoided by a precise volume finder
+  if (m_volumes.external.size()) {
+    for (const auto v : volumes()) {
+      if (v->inside(gctx, nState.position)) {
+        v->updateNavigationState(gctx, nState);
+        return;
+      }
+    }
+  }
+  // The state updator of the volume itself
+  nState.currentVolume = this;
+  m_SurfaceCandidatesUpdator.delegate(gctx, nState);
+  nState.surfaceCandidate = nState.surfaceCandidates.begin();
+  return;
+}
+
+void Acts::Experimental::DetectorVolume::assignSurfaceCandidatesUpdator(
+    ManagedSurfaceCandidatesUpdator&& navStateUpdator,
+    const std::vector<std::shared_ptr<Surface>>& surfaces,
+    const std::vector<std::shared_ptr<DetectorVolume>>& volumes) {
+  m_SurfaceCandidatesUpdator = std::move(navStateUpdator);
+  m_surfaces = ObjectStore<std::shared_ptr<Surface>>(surfaces);
+  m_volumes = ObjectStore<std::shared_ptr<DetectorVolume>>(volumes);
+
+  // This can take a while for complicated geometries
+  // it also needs a geometry context, needs to be figured out
+  // if (not checkContainment(GeometryContext)) {
+  //  throw std::runtime_error("Objects are not contained by volume.");
+  //}
+}
+
+Acts::Extent Acts::Experimental::DetectorVolume::extent(
+    const GeometryContext& gctx, size_t nseg) const {
+  Extent volumeExtent;
+  for (const auto* p : portals()) {
+    volumeExtent.extend(
+        p->surface().polyhedronRepresentation(gctx, nseg).extent());
+  }
+  return volumeExtent;
+}
+
+bool Acts::Experimental::DetectorVolume::checkContainment(
+    const GeometryContext& gctx, size_t nseg) const {
+  // Create the volume extent
+  auto volumeExtent = extent(gctx, nseg);
+  // Check surfaces
+  for (const auto* s : surfaces()) {
+    auto sExtent = s->polyhedronRepresentation(gctx, nseg).extent();
+    if (not volumeExtent.contains(sExtent)) {
+      return false;
+    }
+  }
+  // Check volumes
+  for (const auto* v : volumes()) {
+    auto vExtent = v->extent(gctx, nseg);
+    if (not volumeExtent.contains(vExtent)) {
+      return false;
+    }
+  }
+  // All contained
+  return true;
+}
+
+void Acts::Experimental::DetectorVolume::lock(
+    const GeometryIdentifier& geometryId) {
+  m_geometryId = geometryId;
+
+  // Assign the boundary Identifier
+  GeometryIdentifier portalId = geometryId;
+  for (auto [i, p] : enumerate(m_portals.internal)) {
+    portalId.setBoundary(i + 1);
+    p->assignGeometryId(portalId);
+  }
+
+  // Assign the sensitive/surface Identifier
+  GeometryIdentifier sensitiveId = geometryId;
+  /// @todo add passive count
```

Comment:
important todo?

Reply:
I actually think I will delegate the `GeometryIdentifier` setting entirely, so we can customise it.

Reply:
This will come in a follow-up PR though.